aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -143,50 +143,82 @@ static void check_1d(const Tensor& t, const char* arg, const char* fn) {
 }
 
 Tensor addr(const Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
-  check_1d(vec1, "vec1", "addr");
-  check_1d(vec2, "vec2", "addr");
-  Tensor b_self;
-  std::tie(b_self) = expand_size(self, {vec1.size(0), vec2.size(0)}, "addr");
-  return at::_addr(b_self, vec1, vec2, beta, alpha);
+  Tensor result = at::empty({0}, self.options());
+  at::addr_out(result, self, vec1, vec2, beta, alpha);
+  return result;
 }
 
 Tensor& addr_(Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
-  check_1d(vec1, "vec1", "addr");
-  check_1d(vec2, "vec2", "addr");
-  return at::_addr_(self, vec1, vec2, beta, alpha);
+  return at::addr_out(self, self, vec1, vec2, beta, alpha);
 }
 
 Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
```

Comment:
For `addr_out`, could it be implemented as:

```
auto addr_result = at::addr(self, vec1, vec2, beta, alpha);

// Validates safe casting
const auto result_dtype = addr_result.scalar_type()
TORCH_CHECK(canCast(result_dtype, result.scalar_type()),
  "result type ", result_dtype, " can't be cast to the "
  "desired output type ", result.scalar_type());

at::native::resize_output(result, addr_result.sizes().vec());
result.copy_(addr_result);
return result;
```

This inverts the typical pattern of trying to reuse the "out" variant of a function, but reusing the "out" variant would be very tricky when implementing addr as a composite op. This also suffers because it has to perform a copy from the computed output to the given result, but since we don't care so much about addr's perf I don't think this will be a problem.

Reply:
Thanks for the detailed explanations, this really helps to support type promotion and simplify the code.
I have updated the code accrodingly.